File: src/main/java/io/github/shaksternano/mediamanipulator/logging/DiscordLogger.java
package io.github.shaksternano.mediamanipulator.logging;

import io.github.shaksternano.mediamanipulator.util.LimitedStringBuilder;
import io.github.shaksternano.mediamanipulator.util.StringUtil;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.event.Level;

import java.util.regex.Pattern;

public class DiscordLogger extends InterceptLogger {

    private final MessageChannel channel;

    public DiscordLogger(Logger logger, MessageChannel channel) {
        super(logger);

        if (channel == null) {
            throw new IllegalArgumentException("Channel cannot be null!");
        } else {
            this.channel = channel;
        }
    }

    @Override
    protected void intercept(Level level, String message, @Nullable Throwable t, Object... arguments) {
        LimitedStringBuilder builder = new LimitedStringBuilder(2000);
        String messageWithArguments = formatArguments(message, arguments);
        builder.append("**").append(level).append("** - ").append(getName()).append("\n").append(messageWithArguments);

        if (t != null) {
            builder.append("\nStacktrace:\n").append(StringUtil.getStacktrace(t));
        }

        for (String part : builder.getParts()) {
            channel.sendMessage(part).queue();
        }
    }

    private static String formatArguments(String message, Object... arguments) {
        String messageWithArguments = message;

        for (int i = 0; i < arguments.length; i++) {
            String argument = arguments[i].toString();
            messageWithArguments = messageWithArguments.replaceFirst(Pattern.quote("{}"), argument);
            messageWithArguments = messageWithArguments.replaceAll(Pattern.quote("{" + i + "}"), argument);
        }

        return messageWithArguments;
    }
}

File: discord/src/main/kotlin/io/github/shaksternano/borgar/discord/logging/DiscordLogger.kt
package io.github.shaksternano.borgar.discord.logging

import dev.minn.jda.ktx.generics.getChannel
import io.github.shaksternano.borgar.core.baseLogger
import io.github.shaksternano.borgar.core.logging.InterceptLogger
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel
import net.dv8tion.jda.api.utils.FileUpload
import org.slf4j.Logger
import org.slf4j.event.Level
import java.io.PrintWriter
import java.io.StringWriter

private const val BACKTICK_AND_NEWLINE_LENGTH = 8
private const val MAX_MESSAGE_LENGTH = Message.MAX_CONTENT_LENGTH - BACKTICK_AND_NEWLINE_LENGTH

class DiscordLogger(
    delegate: Logger,
    val logChannelId: Long,
    private val jda: JDA,
) : InterceptLogger(delegate) {

    override fun intercept(level: Level, message: String?, t: Throwable?, vararg arguments: Any?) {
        if (message == null) return
        val logChannel = getLogChannel() ?: return
        val messageWithArguments = formatArguments(message, arguments.asList())
        var formattedMessage = "$level - $name"
        if (messageWithArguments.isNotBlank()) {
            formattedMessage += "\n$messageWithArguments"
        }
        if (t != null) {
            val stackTrace = getStacktrace(t)
            formattedMessage += "\n\nStacktrace:\n$stackTrace"
        }
        formattedMessage = formattedMessage.trim()
        val messageAction = if (formattedMessage.length > MAX_MESSAGE_LENGTH) {
            val bytes = formattedMessage.toByteArray()
            logChannel.sendFiles(FileUpload.fromData(bytes, "message.txt"))
        } else {
            logChannel.sendMessage("```\n$formattedMessage\n```")
        }
        messageAction.queue(null) {
            baseLogger.error("Failed to log message to Discord", it)
        }
    }

    private fun getLogChannel(): MessageChannel? =
        jda.getChannel<MessageChannel>(logChannelId)

    private fun formatArguments(message: String, arguments: List<Any?>): String {
        var messageWithArguments = message
        arguments.forEachIndexed { i, argument ->
            val stringArgument = argument.toString()
            messageWithArguments = messageWithArguments.replaceFirst(Regex.fromLiteral("{}"), stringArgument)
            messageWithArguments = messageWithArguments.replace(Regex.fromLiteral("{$i}"), stringArgument)
        }
        return messageWithArguments
    }

    private fun getStacktrace(t: Throwable): String {
        val stringWriter = StringWriter()
        val printWriter = PrintWriter(stringWriter)
        t.printStackTrace(printWriter)
        return stringWriter.toString()
    }
}
